xla/service/reduce_scatter_all_gather_combiner_test.cc

```diff
@@ -0,0 +1,161 @@
+  EXPECT_EQ(CollectiveCount<HloOpcode::kAllGather>(module), 0);
+  EXPECT_EQ(CollectiveCount<HloOpcode::kReduceScatter>(module), 0);
+  EXPECT_EQ(CollectiveCount<HloOpcode::kAllReduce>(module), 2);
+}
+
+TEST_F(ReduceScatterAllGatherCombinerTest, IncompatibleProperties) {
+  absl::string_view hlo_string = R"(
+HloModule ReduceScatter
+
+add {
+  x = bf16[] parameter(0)
+  y = bf16[] parameter(1)
+  ROOT add = bf16[] add(x, y)
+}
+
+ENTRY main {
+param.1 = bf16[8,128,1024]{2,1,0} parameter(0)
+reduce-scatter.1 = bf16[8,64,1024]{2,1,0} reduce-scatter(param.1), channel_id=8, replica_groups={{0,1},{2,3},{4,5},{6,7}}, use_global_device_ids=false, dimensions={1}, to_apply=add
+all-gather.1 = bf16[8,128,1024]{2,1,0} all-gather(reduce-scatter.1), channel_id=5, replica_groups={{0,1},{2,3},{4,5},{6,7}}, dimensions={1}, use_global_device_ids=true
+}
+)";
+
+  TF_ASSERT_OK_AND_ASSIGN(auto module, RunPass(hlo_string,
+                                               /*num_replicas=*/8,
+                                               /*num_partitions=*/1,
+                                               /*expect_change=*/false));
+}
+
+TEST_F(ReduceScatterAllGatherCombinerTest, ReduceScatterWithMoreThanOneUser) {
+  /* The following pattern should stay untouched as the the ReduceScatters
+     have two users
+
+      Param            Param
+        |                |
+    ReduceScatter   ReduceScatter
+    |          \     /        |
+    AllGather    Add      AllGather
+
+  */
+
+  absl::string_view hlo_string = R"(
+HloModule ReduceScatter
+
+add {
+  x = bf16[] parameter(0)
+  y = bf16[] parameter(1)
+  ROOT add = bf16[] add(x, y)
+}
+
+ENTRY main {
+param.1 = bf16[8,128,1024]{2,1,0} parameter(0)
+param.2 = bf16[8,128,1024]{2,1,0} parameter(1)
+reduce-scatter.1 = bf16[8,64,1024]{2,1,0} reduce-scatter(param.1), channel_id=8, replica_groups={{0,1},{2,3},{4,5},{6,7}}, use_global_device_ids=true, dimensions={1}, to_apply=add
+all-gather.1 = bf16[8,128,1024]{2,1,0} all-gather(reduce-scatter.1), channel_id=5, replica_groups={{0,1},{2,3},{4,5},{6,7}}, dimensions={1}, use_global_device_ids=true
+reduce-scatter.2 = bf16[8,64,1024]{2,1,0} reduce-scatter(param.2), channel_id=9, replica_groups={{0,1},{2,3},{4,5},{6,7}}, use_global_device_ids=true, dimensions={1}, to_apply=add
+all-gather.2 = bf16[8,128,1024]{2,1,0} all-gather(reduce-scatter.2), channel_id=5, replica_groups={{0,1},{2,3},{4,5},{6,7}}, dimensions={1}, use_global_device_ids=true
+
+add.1 = bf16[8,128,1024]{2,1,0} add(all-gather.1, all-gather.2)
+add.2 = bf16[8,64,1024]{2,1,0} add(reduce-scatter.1, reduce-scatter.2)
+}
+)";
+
+  TF_ASSERT_OK_AND_ASSIGN(auto module, RunPass(hlo_string,
+                                               /*num_replicas=*/8,
+                                               /*num_partitions=*/1,
+                                               /*expect_change=*/false));
+  EXPECT_EQ(CollectiveCount<HloOpcode::kAllGather>(module), 2);
```

Comment:
If there is no change, not sure if testing counts adds much.

Reply:
please see the last comment: https://github.com/openxla/xla/pull/5334#discussion_r1318062753

Reply:
I think this comment is still valid. If we don't expect a change, no need to count op counts.